### ISSUE TITLE
Server failing to start while setting for Redis Semantic Cache [Bug]: 

### ISSUE THREAD

**Issue by Soumen96SC** [BerriAI/litellm#3056]
### What happened?

Litellm proxy server is crashing while trying to deploy on the kubernetes cluster.

### Relevant log output

```shell
Getting the following error while deploying the proxy server on kubernetes cluster with semantic cache set as true.

INFO: Started server process [1]
INFO: Waiting for application startup.
ERROR: Traceback (most recent call last):
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 732, in lifespan
async with self.lifespan_context(app) as maybe_state:
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 608, in aenter
await self._router.startup()
File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 709, in startup
await handler()
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 3061, in startup_event
await initialize(**worker_config)
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2905, in initialize
) = await proxy_config.load_config(router=llm_router, config_file_path=config)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2059, in load_config
self._init_cache(cache_params=cache_params)
File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1960, in _init_cache
litellm.cache = Cache(**cache_params)
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 1326, in init
self.cache = RedisSemanticCache(
^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 582, in init
from redisvl.index import SearchIndex
ModuleNotFoundError: No module named 'redisvl'

ERROR: Application startup failed. Exiting.

#------------------------------------------------------------#

'This feature doesn't meet my needs because...'
https://github.com/BerriAI/litellm/issues/new
#------------------------------------------------------------#
```

**Comment by tan-yong-sheng** [comment]
Any workaround? I faced similar issue as follows:

```
Prisma schema loaded from schema.prisma
Datasource "client": PostgreSQL database "postgres", schema "public" at "litellm_db:5432"

The database is already in sync with the Prisma schema.

Running generate... (Use --skip-generate to skip the generators)
root@instance-20240630-1207:~/openwebui#
root@instance-20240630-1207:~/openwebui# docker logs litellm
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_max_budget" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_id" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_name" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_info" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_spend" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_aliases" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_group" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
Installing Prisma CLI
 * Install prebuilt node (22.5.1) ..... done.
Prisma schema loaded from schema.prisma
Datasource "client": PostgreSQL database "postgres", schema "public" at "litellm_db:5432"

The database is already in sync with the Prisma schema.

Running generate... - Prisma Client Python (v0.11.0)

Some types are disabled by default due to being incompatible with Mypy, it is highly recommended
to use Pyright instead and configure Prisma Python to use recursive types. To re-enable certain types:

generator client {
  provider             = "prisma-client-py"
  recursive_type_depth = -1
}

If you need to use Mypy, you can also disable this message by explicitly setting the default value:

generator client {
  provider             = "prisma-client-py"
  recursive_type_depth = 5
}

For more information see: https://prisma-client-py.readthedocs.io/en/stable/reference/limitations/#default-type-limitations

✔ Generated Prisma Client Python (v0.11.0) to ./../../prisma in 247ms

INFO:     Started server process [1]
INFO:     Waiting for application startup.
root@instance-20240630-1207:~/openwebui#
root@instance-20240630-1207:~/openwebui# docker logs litellm
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_max_budget" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_id" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_name" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_info" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_spend" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_aliases" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:160: UserWarning: Field "model_group" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
Installing Prisma CLI
 * Install prebuilt node (22.5.1) ..... done.
Prisma schema loaded from schema.prisma
Datasource "client": PostgreSQL database "postgres", schema "public" at "litellm_db:5432"

The database is already in sync with the Prisma schema.

Running generate... - Prisma Client Python (v0.11.0)

Some types are disabled by default due to being incompatible with Mypy, it is highly recommended
to use Pyright instead and configure Prisma Python to use recursive types. To re-enable certain types:

generator client {
  provider             = "prisma-client-py"
  recursive_type_depth = -1
}

If you need to use Mypy, you can also disable this message by explicitly setting the default value:

generator client {
  provider             = "prisma-client-py"
  recursive_type_depth = 5
}

For more information see: https://prisma-client-py.readthedocs.io/en/stable/reference/limitations/#default-type-limitations

✔ Generated Prisma Client Python (v0.11.0) to ./../../prisma in 247ms

INFO:     Started server process [1]
INFO:     Waiting for application startup.
ERROR:    Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 732, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 608, in __aenter__
    await self._router.startup()
  File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 709, in startup
    await handler()
  File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2572, in startup_event
    await initialize(**worker_config)
  File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 2287, in initialize
    ) = await proxy_config.load_config(router=llm_router, config_file_path=config)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1482, in load_config
    self._init_cache(cache_params=cache_params)
  File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1370, in _init_cache
    litellm.cache = Cache(**cache_params)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 1751, in __init__
    self.cache = RedisSemanticCache(
                 ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litellm/caching.py", line 917, in __init__
    from redisvl.index import SearchIndex
ModuleNotFoundError: No module named 'redisvl'

ERROR:    Application startup failed. Exiting.

#------------------------------------------------------------#
#                                                            #
#       'This feature doesn't meet my needs because...'       #
#        https://github.com/BerriAI/litellm/issues/new        #
#                                                            #
#------------------------------------------------------------#

 Thank you for using LiteLLM! - Krrish & Ishaan



Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new

```

**Comment by mirodrr2** [comment]
Facing the same issue here on `ghcr.io/berriai/litellm:main-latest`

```

INFO: Started server process [1] | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
-- | --
  | 2024-12-10T18:43:14.713Z | INFO: Waiting for application startup. | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ERROR: Traceback (most recent call last): | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 732, in lifespan | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | async with self.lifespan_context(app) as maybe_state: | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 608, in __aenter__ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | await self._router.startup() | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 709, in startup | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | await handler() | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 3119, in startup_event | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ) = await proxy_config.load_config( | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1664, in load_config | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | self._init_cache(cache_params=cache_params) | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/litellm/proxy/proxy_server.py", line 1498, in _init_cache | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | litellm.cache = Cache(**cache_params) | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ^^^^^^^^^^^^^^^^^^^^^ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/litellm/caching/caching.py", line 174, in __init__ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | self.cache = RedisSemanticCache( | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ^^^^^^^^^^^^^^^^^^^ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | File "/usr/local/lib/python3.11/site-packages/litellm/caching/redis_semantic_cache.py", line 34, in __init__ | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z    from redisvl.index import SearchIndex | from redisvl.index import SearchIndex | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156ZModuleNotFoundError: No module named 'redisvl' | ModuleNotFoundError: No module named 'redisvl' | LiteLLM/LiteLLMContainer/8bb941f885524dd28b4484dbb91dd7f9
  | 2024-12-10T18:43:22.156Z | ERROR: Application startup failed. Exiting.
```

**Comment by mirodrr2** [comment]
@tylerhutcherson

Should this be fixed by your PR? 

Did this change ever make it into main? I see it in [BerriAI:litellm_stable_pr_merges](https://github.com/BerriAI/litellm/tree/litellm_stable_pr_merges) but can't see the changes in main

**Comment by mirodrr2** [comment]
@tan-yong-sheng, did you ever get this working?

**Comment by tylerhutcherson** [comment]
@mirodrr2 this looks like `redisvl` is not getting installed on the server? This PR was merged by the team in August: https://github.com/BerriAI/litellm/pull/5412

**Comment by tylerhutcherson** [comment]
It was added to `requirements.txt` in this commit: https://github.com/tylerhutcherson/litellm/blob/77d9d09d4afefed6ea86b506cad140a2a3bfd024/requirements.txt

Is there somewhere else it's missing? I'm not familiar enough with the litellm packaging to know if we are missing something here. Maybe @ishaan-jaff can comment?

**Comment by mirodrr2** [comment]
It looks like it was merged into [BerriAI:litellm_stable_pr_merges](https://github.com/BerriAI/litellm/tree/litellm_stable_pr_merges) but can't see the changes in main

This is the requirements.txt in main https://github.com/tylerhutcherson/litellm/blob/main/requirements.txt, which is missing your `redisvl==0.3.3` addition
@tylerhutcherson

**Comment by mirodrr2** [comment]
Very strange that the changes in [BerriAI:litellm_stable_pr_merges](https://github.com/BerriAI/litellm/tree/litellm_stable_pr_merges) haven't made it to main yet? 

Not sure what that branch is for. Your PR was merged 4 months ago so it should ideally be in main by now...

 @ishaan-jaff please advise how we can get this into main

**Comment by tylerhutcherson** [comment]
Agreed, this is strange. I wasn't aware of a missing step here. Or maybe it was by design. Happy to help rectify.

**Comment by ishaan-jaff** [comment]
ack on the issue, currently working on refactoring our repo / improving code quality. Will look into this issue once that is done 

please ping me if I don't respond on this issue in a week @mirodrr2

**Comment by mirodrr2** [comment]
Will do, thank you

**Comment by ishaan-jaff** [comment]
the reason the PR by @tylerhutcherson  is not in main, is because it failed our testing, but will look at it on our side and investigate

**Comment by mirodrr2** [comment]
@ishaan-jaff , just pinging you since it's been a week

**Comment by mirodrr2** [comment]
@ishaan-jaff, pinging you since it's been (almost) another week. Just hoping for an ack I know you guys are busy

**Comment by mirodrr2** [comment]
@ishaan-jaff pinging since it's been another week

@krrishdholakia also pinging you because I haven't gotten a response from ishaan-jaff in a while

Just hoping for a reply. If other stuff is higher priority I get it, but hoping to have this feature working soon if possible

**Comment by ghost** [comment]
thanks for the bump @mirodrr2 when do you need this by? @mirodrr2

**Comment by mirodrr2** [comment]
I just have some users that expressed interest in trying out semantic caching and I hit this issue while testing it out

Ideally want to be able to tell them it'll be fixed in a week or two when I talk with them tomorrow.

**Comment by ghost** [comment]
Yes - that's a reasonable timeline. I'll try and work on this now

**Comment by mirodrr2** [comment]
@krrishdholakia, any chance this could get prioritized?

**Comment by tylerhutcherson** [comment]
Chiming in as we have 1-2 users interested in LiteLLM + Redis architecture and I was hoping to mock up an example using the integration.

**Comment by tylerhutcherson** [comment]
> the reason the PR by [@tylerhutcherson](https://github.com/tylerhutcherson) is not in main, is because it failed our testing, but will look at it on our side and investigate

And anything we can do related to testing improvements, happy to help. Can also jump on a call to knock this out with you sometime this week.

**Comment by mnotari** [comment]
hi, i'm having the same issue, _**ModuleNotFoundError: No module named 'redisvl'**_ , have you any updates on this? Thanks ! 

> the reason the PR by [@tylerhutcherson](https://github.com/tylerhutcherson) is not in main, is because it failed our testing, but will look at it on our side and investigate
